common/quotas/interfaces.go

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package quotas
+
+import "time"
+
+// Policy corresponds to a quota policy. A policy allows implementing layered
+// and more complex rate limiting functionality.
+type Policy interface {
+	// Allow attempts to allow a request to go through. The method returns
+	// immediately with a true or false indicating if the request can make
+	// progress
+	Allow() bool
+
+	// Wait will wait upto duration d for the request to go through. The method
+	// returns a true or false indicating if the request can make progress.
+	Wait(d time.Duration) bool
```

Comment:
Looks like this method is not being used yet in this PR. I don't feel like this is a good signature that you will use in the future. Venkat already have something similar here: https://github.com/uber/cadence/blob/master/service/matching/ratelimiter.go#L92
Maybe you can do the same, or remove it until when you are going to use it.

Reply:
I'll remove it for now and bring it back if necessary.. For the API quota case I do not think that API is necessary.